timeplus-native-jdbc/src/main/java/com/timeplus/data/type/complex/DataTypeLowCardinality.java

```diff
@@ -104,12 +106,46 @@ public Object deserializeBinary(BinaryDeserializer deserializer) throws SQLExcep
 
     @Override
     public Object[] deserializeBinaryBulk(int rows, BinaryDeserializer deserializer) throws SQLException, IOException {
-        Object[] data = this.nestedDataType.deserializeBinaryBulk(rows, deserializer);
-        return data;
+        if (rows==0) {
+            Object[] data = nestedDataType.deserializeBinaryBulk(rows, deserializer);
+            return data;
+        }
+        else {
+            Long version = deserializer.readLong(); 
+            if (version != this.version) {
+                throw new SQLException("version error in type low_cardinality");
+            }
+            Long index_type = deserializer.readLong() & 0b11111111; //  decide what type it is
+            Long key_nums = deserializer.readLong();
+            Object[] dictionary = nestedDataType.deserializeBinaryBulk(key_nums.intValue(), deserializer);
+            Long row_nums = deserializer.readLong();
+            Object[] data = new Object[rows];
+            for (int i = 0; i < row_nums; i++) {
+                if (index_type == IndexType.UInt8.getValue()) {
+                    data[i]=dictionary[(short) (deserializer.readByte() & 0xff)];
+                }
+                else if (index_type == IndexType.UInt16.getValue()) {
+                    data[i]=dictionary[deserializer.readShort()];
+                }
+                else if (index_type == IndexType.UInt32.getValue()) {
+                    data[i]=dictionary[deserializer.readInt()];
+                }
+                else {
+                    data[i]=dictionary[(int) deserializer.readLong()];
+                }
+                
+            }
```

Comment:
a better way to avoid compare index_type more times:
```java
DataType type;
if (index_type == uint8)
    type = new ... DataTypeUInt8 (or DataTypeFactory.get("uint8");
else (index_type == uint16)
  ...

data = type.deserializeBinaryBulk(rows);
```